aospy/reductions.py

```diff
@@ -0,0 +1,137 @@
+    else:
+        cls.instances[label] = reduction
+
+
+class Reduction(object):
+    """Base class for spatiotemporal reductions."""
+    # TODO: make `instances` read-only except for from the
+    # `_register_reduction` call within `__init__`
+    instances = {}
+
+    @classmethod
+    def _register_reduction(cls, reduction, label):
+        _generic_register_reduction(cls, reduction, label)
+
+    def __init__(self, reduction_func, label):
+        self._register_reduction(self, label)
+        self._reducer = reduction_func
+        if isinstance(label, str):
+            self._label = label
+        else:
+            raise TypeError("'label' must be a string; got type "
+                            "{}".format(label))
+
+    @property
+    def reduce(self):
+        return self._reducer
+
+    @reduce.setter
+    def reduce(self, value):
+        raise ValueError("'reduce' method cannot be modified after Reduction "
+                         "object has been created.")
+
+    @property
+    def label(self):
+        return self._label
+
+    @label.setter
+    def label(self, value):
+        raise ValueError("'label' method cannot be modified after Reduction "
+                         "object has been created.")
+
+    def __call__(self, data, *args, **kwargs):
+        return self.reduce(data, *args, **kwargs)
+
+
+class GriddedReduction(Reduction):
+    # TODO: I think I am overwriting the parent class's `instances` attribute.
+    # How can I register the instances of each subclass?  Or is this fine,
+    # since the generic Reduction class won't ever be used, and other Reduction
+    # subclasses won't overwrite each other's `instances` dictionary, just
+    # the (empty) parent one.
+    instances = {}
+
+    @classmethod
+    def _register_reduction(cls, reduction, label):
+        _generic_register_reduction(cls, reduction, label)
+
+
+class RegionReduction(Reduction):
+    instances = {}
+
+    @classmethod
+    def _register_reduction(cls, reduction, label):
+        _generic_register_reduction(cls, reduction, 'reg.' + label)
+
+    def __init__(self, reduction_func, label, coord_reduction=None):
+        self._register_reduction(self, label)
+        self._reducer = reduction_func
+        self._coord_reducer = coord_reduction
+
+        if isinstance(label, str):
+            self._label = label
+        else:
+            raise TypeError("'label' must be a string; got type "
+                            "{}".format(label))
+
+    @property
+    def label(self):
+        return 'reg.' + self._label
+
+    @label.setter
+    def label(self, value):
+        raise ValueError("'label' method cannot be modified after Reduction "
+                         "object has been created.")
+
+    def __call__(self, data, region, *args, coord_name=None, coord_data=None,
+                 **kwargs):
+        data_out = self.reduce(region.ts(data))
+        if self._coord_reducer is not None and coord_name is not None:
```

Comment:
Could it be more transparent here to have a single keyword argument called `coord` (default `None`) which would accept either a string or a DataArray?  Then the logic here could look like:
```python
if self._coord_reducer is not None and coord is not None:
    if isinstance(coord, str):
        coord_name = coord
        coord_data = data[coord_name]
    elif isinstance(coord, xr.DataArray):
        coord_name = coord.name
        coord_data = coord
    else:
        raise ValueError("'coord' must either be of type str or DataArray")
    data_out[coord_name] = self._coord_reducer(coord_data, *args, **kwargs)
```

Reply:
Yes I like your idea.